server/base/src/main/java/org/apache/accumulo/server/security/SecurityOperation.java

```diff
@@ -106,7 +106,7 @@ public static PermissionHandler getPermHandler(ServerContext context) {
   protected SecurityOperation(ServerContext context, Authorizor author, Authenticator authent,
       PermissionHandler pm) {
     this.context = context;
-    zkUserPath = Constants.ZROOT + "/" + context.getInstanceID() + "/users";
+    zkUserPath = Constants.ZROOT + "/" + context.getInstanceID() + Constants.ZUSERS;
```

Comment:
Every time we use this constant, we are always appending it this way. This implies it might make more sense to just make a method dangle off of `context` that uses a memoized supplier:

```java
    zkUserPath = context.zkUserPath();
```

```java
    Supplier<String> zkUserPath = Suppliers.memoize(() -> Constants.ZROOT + "/" + getInstanceID() + Constants.ZUSERS);
    
    // ...
    
    public String zkUserPath() {
      return zkUserPath.get();
    }
```

Reply:
Maybe as a follow-on?  If I recall, almost always these path fragments are concatenated to build a path and the actual  component name is not needed in isolation.  The issue is that the instance id is a runtime value - but we could migrate to a set of static methods that accepted an instance id and returned a fully formed path.

The start usually is built with `ZooUtil.getRoot(iid)` and then a bunch of concatenations that may be easier to work with if it was  along the lines of `ZooUtil.getZkUserPath(iid, userName)`

It may be preferable to commit this as it maintains the current convention and the use follow-on(s) to reduce the usage of Constant values everywhere.  Or, maybe memorize then on context?  

Not sure which approach would be best, just providing a suggestion for discussion.

Reply:
A good end goal would be to have our ZK connections chroot'd, so we don't have to prefix everything with the path to the instanceID. But, in order to achieve that goal, it would be easier if the paths were concatenated in fewer places. I proposed the improvement here, because I know it would serve that end goal, and it basically changes the same lines that are already being modified here.

Reply:
I was just trying to suggest that 1) we reduce concatenation and 2) we localize the builders in a single place.

Reply:
> I was just trying to suggest that 1) we reduce concatenation and 2) we localize the builders in a single place.

My suggestion achieves that, but just for this one path. The other constants or paths may not be as easy to do (may include more variables... like injection of table name, tserver name, etc.). This one is an easy one to do here, since every case is identical, the only variable component is the instance ID, and they all have a context to work with. So rather than focus on swapping out a constant, which is only a tiny, trivial improvement, we can swap out the string literal for a method, which is a slightly bigger improvement, and works towards another larger goal.

Reply:
Added the context method for determining zkUserPath.